File: data-structure-algorithm/206-Reverse_Linked_List.js
/**
 * Reverse a singly linked list.

Example:

Input: 1->2->3->4->5->NULL
Output: 5->4->3->2->1->NULL
Follow up:

A linked list can be reversed either iteratively or recursively. Could you implement both?
 *
 */

 // 循环
var reverseLinkedList = (head) => {
 let x 
 let y
 while (head) {
    y = head.next
    head.next = x
    x = head
    head = y
 }
 return x
}

// 递归
var reverseLinkedList = (head) => {
  if (!head || !head.next) {
    return head
  }
  let newHead = reverseLinkedList(head.next)
  head.next.next = head
  head.next = null
  return newHead
 }

File: data-structure-algorithm/golang/206. Reverse Linked List.go
package ds

/**

Reverse a singly linked list.

Example:

Input: 1->2->3->4->5->NULL
Output: 5->4->3->2->1->NULL
Follow up:

A linked list can be reversed either iteratively or recursively. Could you implement both?


**/

/**
 * Definition for singly-linked list.
 * type ListNode struct {
 *     Val int
 *     Next *ListNode
 * }
 */
type ListNode struct {
	Val  int
	Next *ListNode
}

// iteratively
func reverseList(head *ListNode) *ListNode {
	var x *ListNode
	var y *ListNode
	// var pre  *ListNode
	for head != nil {
		y = head.Next // y 存 后面的一个
		head.Next = x // x 为 前一个
		x = head      // 替换前一个的指针
		head = y
	}
	return x
}

// recursively
func reverseList2(head *ListNode) *ListNode {
	// func get(pre *ListNode, next *ListNode) {
	// }
	if head == nil || head.Next == nil {
		return head
	}
	newHead := reverseList(head.Next)
	head.Next.Next = head
	head.Next = nil
	return newHead
}
